src/Psl/Type/Exception/AssertException.php

```diff
@@ -17,18 +17,28 @@ final class AssertException extends Exception
     /**
      * @param list<string> $paths
      */
-    private function __construct(string $actual, string $expected, array $paths = [], ?Throwable $previous = null)
+    private function __construct(?string $actual, string $expected, array $paths = [], ?Throwable $previous = null)
     {
         $first = $previous instanceof Exception ? $previous->getFirstFailingActualType() : $actual;
 
-        parent::__construct(
-            Str\format(
+        if ($first !== null) {
+            $message = Str\format(
                 'Expected "%s", got "%s"%s.',
                 $expected,
                 $first,
-                $paths ? ' at path "' . Str\join($paths, '.') . '"' : ''
-            ),
-            $actual,
+                $paths ? ' at path "' . Str\join($paths, '.') . '"' : '',
+            );
+        } else {
+            $message = Str\format(
```

Comment:
It might make more sense to build the messages from the named constructors instead of performing if/else logic here?

Reply:
Could be, yes, I tried to make as few changes as possible - as the primary goal at this stage is to build some common grounds :-)

Reply:
Okay, done, will keep the history for now, and will squash if the PR is ever ready to be accepted.